File: app_codebase/python_base_05/bin/core/ws/response/ws_response.py
"""Encode/decode WebSocket JSON frames using the HTTP JSON envelope."""

from __future__ import annotations

import json
from typing import Any

from core.http.response.response import json_error_body, json_success_body


def encode_frame(*, ok: bool, data: object | None = None, code: str = "", message: str = "") -> str:
    if ok:
        body = json_success_body(data)
    else:
        body = json_error_body(code=code, message=message)
    return json.dumps(body)


def encode_ok(data: object | None) -> str:
    return encode_frame(ok=True, data=data)


def encode_error(*, code: str, message: str) -> str:
    return encode_frame(ok=False, code=code, message=message)


def parse_incoming(text: str) -> tuple[dict[str, Any] | None, dict[str, Any] | None]:
    """Return (data dict, error dict) — one will be set."""
    try:
        raw = json.loads(text)
    except json.JSONDecodeError:
        return None, {"code": "invalid_json", "message": "Message must be valid JSON"}
    if not isinstance(raw, dict):
        return None, {"code": "invalid_json", "message": "Message must be a JSON object"}
    if raw.get("ok") is True and isinstance(raw.get("data"), dict):
        return raw["data"], None
    if raw.get("ok") is False and isinstance(raw.get("error"), dict):
        err = raw["error"]
        return None, {
            "code": str(err.get("code", "invalid_message")),
            "message": str(err.get("message", "Invalid message")),
        }
    if "type" in raw and "channel" in raw:
        return raw, None
    return None, {"code": "invalid_message", "message": "Expected {type, channel} in data"}

File: app_codebase/dart_bkend_base_02/bin/core/ws/response/ws_response.dart
/// Encode/decode WebSocket JSON frames using the HTTP JSON envelope.
library;

import 'dart:convert';

import '../../http/response/response.dart';

String encodeWsOk(Object? data) => jsonEncode(jsonSuccessBody(data));

String encodeWsError({required String code, required String message}) =>
    jsonEncode(jsonErrorBody(code: code, message: message));

({Map<String, dynamic>? data, Map<String, String>? error}) parseIncoming(
  String text,
) {
  try {
    final raw = jsonDecode(text);
    if (raw is! Map) {
      return (
        data: null,
        error: {'code': 'invalid_json', 'message': 'Message must be a JSON object'},
      );
    }
    final map = Map<String, dynamic>.from(raw);
    if (map['ok'] == true && map['data'] is Map) {
      return (data: Map<String, dynamic>.from(map['data'] as Map), error: null);
    }
    if (map['ok'] == false && map['error'] is Map) {
      final err = Map<String, dynamic>.from(map['error'] as Map);
      return (
        data: null,
        error: {
          'code': err['code']?.toString() ?? 'invalid_message',
          'message': err['message']?.toString() ?? 'Invalid message',
        },
      );
    }
    if (map.containsKey('type') && map.containsKey('channel')) {
      return (data: map, error: null);
    }
    return (
      data: null,
      error: {
        'code': 'invalid_message',
        'message': 'Expected {type, channel} in data',
      },
    );
  } catch (_) {
    return (
      data: null,
      error: {'code': 'invalid_json', 'message': 'Message must be valid JSON'},
    );
  }
}
